File: code/strategy/strategy.clj
(ns strategy-clj)

;; A function maker (higher order function)... returns an archive function depending on the strategy
(defn archive [extension archive-fn files archive-name]
  (str (archive-fn files) ", " archive-name "." extension))

(def archive-to-zip (partial archive "zip" (fn [files] "compressed to zip")))
(def archive-to-tar (partial archive "tar" (fn [files] "archived to tar")))

(defn demo
  [& args]
  (println (archive-to-zip ["file1" "file2" "file3"] "zipfile"))
  (println (archive-to-tar ["file1" "file2" "file3"] "tarfile")))

File: code/strategy/strategy.hs
type File = String
type Extension = String
type Name = String

archive ::  Extension -> ([File] -> File) -> [File] -> Name -> File
archive ext f files name = (f files) ++ ", " ++ name ++ "." ++ ext

main :: IO ()
main = do
  putStrLn (archiveToZip ["one", "two"] "zipfile")
  putStrLn (archiveToTar ["one", "two"] "tarfile")
  
  where
    archiveToZip :: [File] -> String -> File
    archiveToZip = archive "zip" (\_ -> "compressed to zip")
    
    archiveToTar :: [File] -> String -> File
    archiveToTar = archive "tar" (\_ -> "archived to tar")